monailabel/scribbles/utils.py

```diff
@@ -160,3 +165,115 @@ def make_likelihood_image_histogram(image, scrib, scribbles_bg_label, scribbles_
+    # based on https://github.com/jiviteshjain/grabcut/blob/main/src/grabcut.ipynb
+    bg_gmm, fg_gmm = sklearn_fit_gmm(image, scrib, scribbles_bg_label, scribbles_fg_label, num_mixtures)
+
+    # add empty channel to image and scrib to be inline with pytorch layout
+    bg_prob = bg_gmm.score_samples(image.reshape((-1, 1))).reshape(image.shape).astype(np.float32)
+    fg_prob = fg_gmm.score_samples(image.reshape((-1, 1))).reshape(image.shape).astype(np.float32)
+
+    bg_prob = np.exp(bg_prob)
+    fg_prob = np.exp(fg_prob)
+
+    gmm_output = np.concatenate([bg_prob, fg_prob])
+
+    return gmm_output
+
+
+def learn_and_apply_gmm_monai(image, scrib, scribbles_bg_label, scribbles_fg_label, num_mixtures):
+    # this function is limited to binary segmentation at the moment
+    n_classes = 2
+
+    # make trimap
+    trimap = np.zeros_like(scrib).astype(np.int32)
+
+    # fetch anything that is not scribbles
+    not_scribbles = ~((scrib == scribbles_bg_label) | (scrib == scribbles_fg_label))
+
+    # set these to 0 == unused
+    trimap[not_scribbles] = -1
+
+    # set background scrib to -1
+    trimap[scrib == scribbles_bg_label] = 0
+    # set foreground scrib to 1
+    trimap[scrib == scribbles_fg_label] = 1
+
+    # add empty channel to image and scrib to be inline with pytorch layout
+    image = np.expand_dims(image, axis=0)
+    trimap = np.expand_dims(trimap, axis=0)
+
+    # transfer everything to pytorch tensor,
+    # we only use CUDA as GMM from MONAI is only available on CUDA atm (2021/04/29)
+    # if no cuda device found, then exit now
+    if not torch.cuda.is_available():
+        raise OSError("Unable to find CUDA device, check your torch/monai installation")
+
+    device = "cuda"
+    image = torch.from_numpy(image).type(torch.float32).to(device)
+    trimap = torch.from_numpy(trimap).type(torch.int32).to(device)
+
+    # initialise our GMM
+    gmm = GaussianMixtureModel(
+        image.size(1),
+        mixture_count=n_classes,
+        mixture_size=num_mixtures,
+        verbose_build=False,
+    )
+    # gmm.reset()
+
+    # learn gmm from image and trimap
+    gmm.learn(image, trimap)
+
+    # apply gmm on image
+    gmm_output = gmm.apply(image)
+
+    return gmm_output.squeeze(0).cpu().numpy()
+
+
+def make_likelihood_image_gmm(
+    image,
+    scrib,
+    scribbles_bg_label,
+    scribbles_fg_label,
+    num_mixtures=20,
+    return_label=False,
+):
+    # learn gmm and apply to image, return output label prob
+    try:
+        # this may fail if MONAI Cpp Extensions are not loaded properly
+        retprob = learn_and_apply_gmm_monai(
+            image=image,
+            scrib=scrib,
+            scribbles_bg_label=scribbles_bg_label,
+            scribbles_fg_label=scribbles_fg_label,
+            num_mixtures=num_mixtures,
+        )
+    except:
+        # if MONAI's GMM fails, fallback to scikit-learn's GMM
```

Comment:
It would be better to have a MONAI-based fallback. If the issue is that the current MONAI GMM is a cuda or c++ extension that may not be available, maybe we should have a second python-only fallback in MONAI. An implementation for it can for example be borrowed from https://github.com/ldeecke/gmm-torch

Reply:
This fallback is because GMM in MONAI only supports CUDA tensors:
https://github.com/Project-MONAI/MONAI/blob/13314320c4c85e2f3da50a71e3d8ec54befe7cd7/monai/_extensions/gmm/gmm_cpu.cpp#L25

Thanks for pointing to gmm-torch library. If I have bandwidth, I will look at this. I also have a OpenCV-GMM that I did back in my uni days :see_no_evil: (https://github.com/masadcv/OpenCVGMM) - not entirely sure if this is accurate or up-to-date. Will check..

Reply:
Right, I had not looked in the `gmm_cpu.cpp` file. It's indeed more than a packaging issue.

Reply:
There may still be a packaging issue with this on Windows - like we had for SimpleCRF - as JIT would need to compile the source when MONAI GMM is called...